### PR TITLE
Hilbert + unique_id sorting

### PR DESCRIPTION
**Pull request by roystgnr**
If this works, it should enable slit meshes, overset meshes, and all the other fun meshes where nodes or centroids share the same location and therefore the same Hilbert indices.

This is starting to pass my own tests but is not ready to merge; even if it actually passes MooseBuild already I've still got more testing lined up.

### CODE REVIEW

**Comment by dknez** [comment]
Very cool. I can test this on some slit meshes soon, when I get some time. Maybe I'll wait until you've finished your first round of testing though, if that's OK.

**Comment by roystgnr** [comment]
Not just OK, but a very good strategy, in hindsight.  ;-)

I'm pretty baffled by those results, though.  The MOOSE Nemesis tests are the only things failing in debug mode, but there's dozens of crashes in opt mode?

I'm planning to get a slit mesh unit test (and maybe an overset mesh test) in there, but running on big complicated slit meshes would be better.

**Comment by dknez** [comment]
OK, no problem.

I can send you a 3D slit mesh now to test on, if you like. Otherwise, I can do that test later, once these moose tests are green.

**Comment by roystgnr** [comment]
We need to tell MooseBuild to run the dbg mode tests in parallel too.  Took me forever to realize that the reason I couldn't replicate failures was nothing more than the fact that I was neglecting to run with "-p 2".

**Comment by jwpeterson** [comment]
> We need to tell MooseBuild to run the dbg mode tests in parallel too.

Hmm... yeah I don't see why not.  I'll open an issue about that.

**Comment by roystgnr** [comment]
Okay, this passes all the libMesh, MOOSE, and GRINS tests I can throw at it.  And it's a good patch - it brings unique_id() support to a hundred corner cases instead of "maybe it'll work if you hold it really steady and don't breathe too hard".

But the *sorting* isn't actually useful yet, because we don't write out nodal unique_id and so we can't use it to disambiguate colocated nodes.  I'll start a separate branch and PR for that bit, and I won't merge this until that's ready to go, in case anyone wants to take time to review it first.

**Comment by permcody** [comment]
Well thanks for working on this.

File: examples/fem_system/fem_system_ex4/fem_system_ex4.C [hunk 1]
```diff
@@ -274,9 +290,8 @@ int main (int argc, char** argv)
 #ifdef LIBMESH_HAVE_EXODUS_API
   // We want to write the file in the ExodusII format, but we don't
```
Comment (jwpeterson) [inline examples/fem_system/fem_system_ex4/fem_system_ex4.C]: Looks like you can delete the comment on lines 291-292 now.

**Comment by jwpeterson** [comment]
I'm pretty sure all the moosebuild targets configure with `--enable-unique-id`.  I don't suppose you tested this branch with `--disable-unique-id`.

File: include/mesh/mesh_communication.h [hunk 2]
```diff
@@ -171,13 +171,20 @@ class MeshCommunication
    */
   void make_elems_parallel_consistent (MeshBase &);
 
-  /**
+   /**
```
Comment (jwpeterson) [inline include/mesh/mesh_communication.h]: Looks like an extra space got inserted here.

**Comment by roystgnr** [comment]
What, do you think I'm an idiot?

(/me sheepishly goes to test with --disable-unique-id)

**Comment by jwpeterson** [comment]
Looks good, did you want to add any unit tests for the new functionality?  I guess they would need to run in parallel to really test it...

**Comment by roystgnr** [comment]
Not sure unit tests for these specific methods would help much; all the bugs which cropped up along the way felt more like interactions between the way different methods exptected things to work.

What we ought to have, though, is unit tests for all the mesh I/O and MeshGeneration options...

**Comment by roystgnr** [comment]
We ought to merge this soon (assuming I didn't screw up any of the conflict resolutions and it still passes tests).  The additional tests I'm working on adding can all go in #765.  With the parallel failure discovered in #766, we should now have a clear case of "previously-broken case which this PR fixes", which is good enough for me.

**Comment by dknez** [comment]
@roystgnr I sent you a test case a while ago, did you get a chance to try this PR on that? If not, let me know and I'll try it.

**Comment by roystgnr** [comment]
Not yet.  If it doesn't require a restart you could try it now; if it does then we'll need to wait until these tests pass and I rebase #765 and then we can try it on that.

**Comment by dknez** [comment]
It doesn't require a restart. It just loads up a 3D mesh with overlapping nodes and checks that there's no ID clash.

**Comment by roystgnr** [comment]
Yeah, that should work with this.  Let me know?  Thanks!

**Comment by dknez** [comment]
I just checked that test case and it worked fine. Thanks for this PR!

**Comment by roystgnr** [comment]
Okay; definitely planning to merge once the debug tests are complete, then.  And I'll put up the rebased version of #765 ASAP.  After this PR is merged, the state of libMesh will be "meshes with collocated dof objects work, until you do a restart with them, at which point dof values on collocated nodes may be silently randomly swapped"... which in hindsight is actually kind of scarier than the simple "meshes with collocated dof objects don't work"
